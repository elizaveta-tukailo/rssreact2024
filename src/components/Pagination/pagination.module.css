.pagination {
  margin: 0 auto 60px;
  width: max-content;
}
.pagination .paginationItem {
  cursor: pointer;
  margin: 0 6px;
  display: inline-block;
  width: 20px;
  height: 20px;
  border-radius: 2px;
  border: 1px solid #2c7dfb;
  text-decoration: none;
  text-align: center;
}
.active {
  color: #ffffff;
  background-color: #2c7dfb;
}
.pagination .paginationItem:hover {
  transition: all 0.3s ease;
  background-color: #2c7dfb;
  color: #ffffff;
}
.pagination.dark .paginationItem {
  background-color: #1e1d1d;
  border-color: #1e1d1d;
  color: #ffffff;
}
.pagination.dark .paginationItem:hover,
.pagination.dark .paginationItem.active:hover {
  opacity: 0.5;
}
.disabledPagination {
  pointer-events: none;
}
.item:hover {
  opacity: 0.5;
  cursor: pointer;
}
.item:active {
  opacity: 0.5;
}
.paginationItem.active {
  opacity: 0.5;
}
