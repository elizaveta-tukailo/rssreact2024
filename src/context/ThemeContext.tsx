import {
  createContext,
  ReactNode,
  useContext,
  useMemo,
  useState,
  useCallback,
} from 'react';
type Theme = 'light' | 'dark';
interface ThemeContextType {
  theme: Theme;
  toggleTheme: () => void;
}
export const ThemeContext = createContext<ThemeContextType | undefined>(
  undefined
);
export function useTheme() {
  const context = useContext(ThemeContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}
interface ThemeProviderProps {
  children: ReactNode;
}
export function ThemeProvider({ children }: ThemeProviderProps) {
  const [theme, setTheme] = useState<Theme>('light');
  const toggleTheme = useCallback(() => {
    setTheme((prevTheme) => (prevTheme === 'light' ? 'dark' : 'light'));
  }, []);
  const value = useMemo(() => ({ theme, toggleTheme }), [theme, toggleTheme]);
  return (
    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
  );
}
