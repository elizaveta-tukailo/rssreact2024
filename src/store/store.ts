import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import { characterApi } from './../services/character';

export const store = configureStore({
  reducer: {
    [characterApi.reducerPath]: characterApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(characterApi.middleware),
});

setupListeners(store.dispatch);
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
