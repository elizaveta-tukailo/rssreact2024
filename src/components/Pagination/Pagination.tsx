import styles from './pagination.module.css';
import { NavLink } from 'react-router-dom';
import { useTheme } from '../../context/ThemeContext';

interface PaginationProps {
  totalCount: number;
  currentPage: number;
}

const Pagination: React.FC<PaginationProps> = ({ totalCount, currentPage }) => {
  const { theme } = useTheme();

  const itemsCount = totalCount;
  const searchQuery = localStorage.getItem('searchQuery');

  const itemsPerPage = 20;
  const pages = [];
  const pagesCount = Math.ceil(itemsCount / itemsPerPage);
  for (let i = 1; i <= pagesCount; i++) {
    pages.push(i);
  }

  const maxPagesToShow = 6;
  let startPage = Math.max(1, currentPage - Math.floor(maxPagesToShow / 2));
  let endPage = startPage + maxPagesToShow - 1;

  if (endPage > pagesCount) {
    endPage = pagesCount;
    startPage = Math.max(1, endPage - maxPagesToShow + 1);
  }

  const paginationsOnPage = pages.slice(startPage - 1, endPage);

  const searchParams = searchQuery ? `?search=${searchQuery}` : '';

  return (
    <div className="container">
      <div className={`${styles.pagination} ${styles[theme]}`}>
        {paginationsOnPage.map((page) => (
          <NavLink
            to={`/page/${page}${searchParams}`}
            className={({ isActive }) => (isActive ? styles.active : undefined)}
            key={page + 1}
          >
            {page}
          </NavLink>
        ))}
      </div>
    </div>
  );
};

export default Pagination;
