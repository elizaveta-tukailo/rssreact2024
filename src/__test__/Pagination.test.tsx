import { render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import Pagination from '../components/Pagination/Pagination';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from '../context/ThemeContext';
import { Provider } from 'react-redux';
import { store } from '../store/store';

describe('Pagination test', () => {
  it('The URL changes when the page changes', async () => {
    render(
      <BrowserRouter>
        <ThemeProvider>
          <Pagination totalCount={80} currentPage={1} />
        </ThemeProvider>
      </BrowserRouter>
    );
    expect(screen.getByText('4'));
    await waitFor(() => {
      userEvent.click(screen.getByText('4'));
      expect(window.location.search).toBe('');
    });
  });
  it('Pagination component renders correctly', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <ThemeProvider>
              <Pagination totalCount={80} currentPage={1} />
            </ThemeProvider>
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
