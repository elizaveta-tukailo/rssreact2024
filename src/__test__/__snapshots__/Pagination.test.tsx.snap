// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination test Pagination component renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="pagination light"
  >
    <a
      href="/page/1"
      onClick={[Function]}
    >
      1
    </a>
    <a
      href="/page/2"
      onClick={[Function]}
    >
      2
    </a>
    <a
      href="/page/3"
      onClick={[Function]}
    >
      3
    </a>
    <a
      href="/page/4"
      onClick={[Function]}
    >
      4
    </a>
  </div>
</div>
`;
